ouroboros-consensus-test/src/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,638 @@
+    --       to be filled in (by the init step above).
+    --
+    --     * If some subsequent window is sparse, then its final slot is filled
+    --       in (by this loop). It must never fill in any older slot in the
+    --       window because those slots have already been sampled (either by
+    --       the init step above or by previous iterations of this loop).
+    --
+    --     * Every slot that was not filled in was drawn from @mbAsc@.
+    --
+    --     * In total: the init step uniformly fills the first window up to
+    --       @numerator@, and then each slot not in the first window is either
+    --       forced to @1@ by its preceding @denominator - 1@ samples or is
+    --       sampled from @mbAsc@.
+    C.forRangeM_ (C.windowSize remainingFullWindows) $ \(C.frWin remainingFullWindows -> islot) -> do
+        -- NB will not be truncated
+        C.SomeWindow Proxy ehcgSlots <- pure $ C.withWindow sz (C.Lbl @EhcgLbl) islot (C.toSize denominator)
+
+        tot <- do
+            tot <- readSTRef rtot
+            end <- BV.testMV S.notInverted mv (C.windowLast ehcgSlots)
+            pure $ (if end then (+1) else id) $ tot
+
+        let sparse = tot == numerator - 1   -- see LOOP INVARIANT
+
+        tot' <- if not sparse then pure tot else do
+            BV.setMV S.notInverted mv (C.windowLast ehcgSlots)
+            pure numerator
+
+        start <- BV.testMV S.notInverted mv (C.windowStart ehcgSlots)
+        writeSTRef rtot $! (if start then subtract 1 else id) $ tot'
+
+    pure mv
+  where
+    UnsafeCheckedHonestRecipe {
+        chrEhcgDensity
+      ,
+        chrWin         = slots
+      }  = recipe
+
+    sz  = C.windowSize slots :: C.Size hon SlotE   -- ie 'Len'
+
+    wrap v = ChainSchedule slots v
+
+-----
+
+-- | The population count of a sample drawn from the stationary distribution of
+-- the idealized honest chain generator
+genFirstWindowPopCount :: R.RandomGen g => Asc -> BV.SomeDensityWindow S.NotInverted -> g -> (Int, g)
+genFirstWindowPopCount asc densityWindow g =
+    let q :: Double
+        (q, g') = R.random g   -- note that q is between 0 and 1.
+
+        cumul = tail $ scanl (+) 0 $ map abs $ Mat.toList sd
+
+        -- A sample drawn from the weighted categorical distribution that is
+        -- the Markov chain's stationary distribution.
+        i = length $ takeWhile (\x -> x / last cumul < q) cumul
+    in
+    (n + i, g')
+  where
+    f = ascVal asc
+
+    BV.SomeDensityWindow (C.Count n) (C.Count d) = densityWindow
+
+    sz = d - n + 1
+
+    dd = toEnum d :: Double
+    nn = toEnum n :: Double
+
+    -- the non-normalized stationary distribution
+    --
+    -- 1) See
+    -- https://stephens999.github.io/fiveMinuteStats/markov_chains_discrete_stationary_dist.html
+    --
+    -- 2) The stationary distribution is the eigenvector with eigenvalue 1.
+    --
+    -- 3) We want the /left/ eigenvector (TODO is this fundamental to transition matrices?).
```

Comment:
> is this fundamental to transition matrices

It is purely a consequence of the convention that the entry $P_{ij}$ in the $i$-th row and the $j$-the column of the transition matrix $P$ is the probability to transition to state $j$ if we are in state $i$ (instead of the opposite), and the way matrix multiplication is defined.